planner/scan.go

```diff
@@ -60,17 +63,84 @@ func (n *scanNode) Kind() string {
 
 func (n *scanNode) Init() error {
 	// init the fetcher
-	if err := n.fetcher.Init(&n.desc, n.fields, n.reverse, n.showDeleted); err != nil {
+	if err := n.fetcher.Init(&n.desc, n.fields, n.filter, n.slct.DocumentMapping, n.reverse, n.showDeleted); err != nil {
 		return err
 	}
 	return n.initScan()
 }
 
 func (n *scanNode) initCollection(desc client.CollectionDescription) error {
 	n.desc = desc
+	n.initFields(n.slct.Fields)
+	if n.slct.OrderBy != nil {
+		for _, cond := range n.slct.OrderBy.Conditions {
+			for _, index := range cond.FieldIndexes {
+				name, _ := n.docMapper.documentMapping.TryToFindNameFromIndex(index)
+				n.tryAddField(name)
+			}
+		}
+	}
 	return nil
 }
 
+func (n *scanNode) initFields(fields []mapper.Requestable) {
+	for _, r := range fields {
+		// add all the possible base level fields the fetcher is responsible
+		// for, including those that are needed by higher level aggregates
+		// or grouping alls, which them selves might have further dependants
+		switch requestable := r.(type) {
+		// field is simple as its just a base level field
+		case *mapper.Field:
+			n.tryAddField(requestable.GetName())
+		// select might have its own select fields and filters fields
+		// @todo: Might need to check that these are sub selects
+		// are _groups
+		case *mapper.Select:
+			if requestable.Targetable.Filter != nil {
+				fieldDescs := parser.ParseFilterFieldsForDescription(
+					requestable.Targetable.Filter.ExternalConditions,
+					n.desc.Schema,
+				)
+				for _, fd := range fieldDescs {
+					n.tryAddField(fd.Name)
+				}
+			}
+			n.tryAddField(requestable.Field.Name + "_id") // foreign key for type joins
+			n.initFields(requestable.Fields)
+		// aggregate might have its own target fields and filter fields
+		case *mapper.Aggregate:
+			for _, target := range requestable.AggregateTargets {
+				if target.Filter != nil {
+					fieldDescs := parser.ParseFilterFieldsForDescription(
+						target.Filter.ExternalConditions,
+						n.desc.Schema,
+					)
+					for _, fd := range fieldDescs {
+						n.tryAddField(fd.Name)
+					}
+				}
+				if target.ChildTarget.HasValue {
+					n.tryAddField(target.ChildTarget.Name)
+				} else {
+					n.tryAddField(target.Field.Name)
+				}
+			}
+		}
+	}
+}
+
+func (n *scanNode) tryAddField(fieldName string) bool {
```

Comment:
todo: As far as I can see the returned boolean is never used, can you please remove it (or use it), and maybe rename this function to `addIfNotExists` or similar